src/daemons/linuxd/src/linuxd/mod.rs

```diff
@@ -68,179 +76,275 @@ pub struct LinuxDaemon {
 
 impl LinuxDaemon {
     pub fn init(
+        control_plane_stream: SocketStream,
         uvm_stream: SocketStream,
         gateway_stream: Option<SocketStream>,
     ) -> Result<Self, Error> {
         Ok(Self {
-            assembler: Arc::new(Mutex::new(RequestAssembler::default())),
+            control_plane_stream,
             uvm_stream,
             gateway_stream,
+            assembler: Arc::new(Mutex::new(RequestAssembler::default())),
             venv: Arc::new(Mutex::new(VirtualEnviromentDirectory::new())),
         })
     }
 
-    pub fn run(self) -> Result<(), Error> {
+    /// Helper method to close a connection to a user VM identified by the connection id. Closing
+    /// the connection also involes stopping all associated worker threads.
+    fn close_connection(
+        uvm_handle: UserVmHandle,
+        poll: &Poll,
+        worker_threads: Option<VecDeque<WorkerThreadHandle>>,
+    ) {
+        // De-register the user VM socket from the poll structure.
+        let user_vm_stream: Arc<Mutex<(SocketStream, VecDeque<u8>)>> =
+            uvm_handle.get_user_vm_stream();
+        match user_vm_stream.lock() {
+            Ok(mut guard) => {
+                let (locked_uvm_stream, _): &mut (SocketStream, VecDeque<u8>) = &mut guard;
+                if let Err(e) = poll.registry().deregister(&mut *locked_uvm_stream) {
+                    error!("failed to de-revister user VM from poll (error={e:?})");
+                }
+            },
+            Err(e) => {
+                error!("error acquiring lock on user VM stream (error={e:?})");
+            },
+        };
+
+        // Send a shutdown message to all worker threads associated
+        // with this user VM.
+        if let Some(mut worker_threads) = worker_threads {
+            for worker_thread in worker_threads.drain(..) {
+                trace!("sending interrupt to worker thread (thread_id={:?})", worker_thread.id);
+
+                // If any of the commands fail, continue trying to drain the remainnig
+                // threads.
+                match worker_thread.cmd_tx.send(VenvCommand::Shutdown) {
+                    Ok(_) => {},
+                    Err(e) => {
+                        error!("error sending shutdown command to worker thread (thread_id={e:?})");
+                        continue;
+                    },
+                }
+                match worker_thread.stop() {
+                    Ok(_) => {},
+                    Err(e) => {
+                        error!("error sending interrupt to worker thread (thread_id={e:?})");
+                        continue;
+                    },
+                }
+                match worker_thread.handle.join() {
+                    Ok(_) => {},
+                    Err(e) => {
+                        error!("error joining worker thread (thread_id={e:?})");
+                    },
+                }
```

Comment:
As we care only about the error paths, switch to the `if let Err(e) = ...` pattern to keep the code easier to read.